beamer/agent.py

```diff
@@ -31,107 +37,165 @@ def _get_deployment_block(contract_info: dict[str, ContractInfo]) -> BlockNumber
     return min(request_manager_deployment_block, fill_manager_deployment_block)
 
 
+@dataclass
+class _BaseChain:
+    w3: Web3
+    id: ChainId
+    rpc_url: URL
+
+
+@dataclass
+class _Chain(_BaseChain):
+    contracts_info: dict[str, ContractInfo]
+    contracts: dict[str, Contract]
+
+    @property
+    def request_manager(self) -> RequestManager:
+        return self.contracts["RequestManager"]
+
+    @property
+    def fill_manager(self) -> FillManager:
+        return self.contracts["FillManager"]
+
+
 class Agent:
     def __init__(self, config: Config):
         self._config = config
         self._stopped = threading.Event()
         self._stopped.set()
         self._init()
 
-    def _init(self) -> None:
-        # Just add one worker, as this effectively serializes the work
-        # This is necessary as we use the account for all resolutions and
-        # would otherwise run into nonce problems
-        self._task_pool = ThreadPoolExecutor(max_workers=1)
-
-        w3_l1 = make_web3(self._config.l1_rpc_url, self._config.account)
-        w3_source = make_web3(self._config.l2a_rpc_url, self._config.account)
-        w3_target = make_web3(self._config.l2b_rpc_url, self._config.account)
-
-        source_chain_id = ChainId(w3_source.eth.chain_id)
-        target_chain_id = ChainId(w3_target.eth.chain_id)
-        chain_ids = {source_chain_id, target_chain_id}
-
-        w3_l1.middleware_onion.add(latest_block_based_cache_middleware)
-
-        source_contracts_info = _get_contracts_info(self._config, w3_source)
-        target_contracts_info = _get_contracts_info(self._config, w3_target)
+    def _init_l1_chain(self) -> _BaseChain:
+        l1_w3 = make_web3(self._config.rpc_urls["l1"], self._config.account)
+        l1_w3.middleware_onion.add(latest_block_based_cache_middleware)
+        chain_id = ChainId(l1_w3.eth.chain_id)
+        return _BaseChain(w3=l1_w3, id=chain_id, rpc_url=self._config.rpc_urls["l1"])
+
+    def _init_chains(self) -> dict[ChainId, _Chain]:
+        chains: dict[ChainId, _Chain] = {}
+        for chain_name, rpc_url in self._config.rpc_urls.items():
+            if chain_name == "l1":
+                continue
+            w3 = make_web3(rpc_url, self._config.account)
+            chain_id = ChainId(w3.eth.chain_id)
+            if chain_id in chains:
+                continue
+            contracts_info = _get_contracts_info(self._config, w3)
```

Comment:
Here we are passing `w3` to `_get_contracts_info`, which uses it just to get the chain ID.
However, we have just obtained the chain ID just a few lines above, so we can, instead of passing a `Web3` instance, give `_get_contracts_info` the chain ID directly and thereby save one RPC call.